src/mcp/server/fastmcp/utilities/func_metadata.py

```diff
@@ -298,58 +303,58 @@ def _try_create_model_and_schema(
     # First handle special case: None
-    if annotation is None:
-        model = _create_wrapped_model(func_name, annotation, field_info)
+    if type_expr is None:
+        model = _create_wrapped_model(func_name, type_expr)
         wrap_output = True
 
     # Handle GenericAlias types (list[str], dict[str, int], Union[str, int], etc.)
-    elif isinstance(annotation, GenericAlias):
-        origin = get_origin(annotation)
+    elif isinstance(type_expr, GenericAlias):
+        origin = get_origin(type_expr)
 
         # Special case: dict with string keys can use RootModel
         if origin is dict:
-            args = get_args(annotation)
+            args = get_args(type_expr)
             if len(args) == 2 and args[0] is str:
-                model = _create_dict_model(func_name, annotation)
+                model = _create_dict_model(func_name, type_expr)
             else:
                 # dict with non-str keys needs wrapping
-                model = _create_wrapped_model(func_name, annotation, field_info)
+                model = _create_wrapped_model(func_name, type_expr)
                 wrap_output = True
         else:
             # All other generic types need wrapping (list, tuple, Union, Optional, etc.)
-            model = _create_wrapped_model(func_name, annotation, field_info)
+            model = _create_wrapped_model(func_name, type_expr)
             wrap_output = True
 
     # Handle regular type objects
-    elif isinstance(annotation, type):
-        type_annotation: type[Any] = cast(type[Any], annotation)
+    elif isinstance(type_expr, type):
+        type_annotation = cast(type[Any], type_expr)
 
         # Case 1: BaseModel subclasses (can be used directly)
-        if issubclass(annotation, BaseModel):
-            model = annotation
+        if issubclass(type_annotation, BaseModel):
+            model = type_annotation
 
-        # Case 2: TypedDict (special dict subclass with __annotations__)
-        elif hasattr(type_annotation, "__annotations__") and issubclass(annotation, dict):
+        # Case 2: TypedDicts:
+        elif is_typeddict(type_annotation):
             model = _create_model_from_typeddict(type_annotation)
 
         # Case 3: Primitive types that need wrapping
-        elif annotation in (str, int, float, bool, bytes, type(None)):
-            model = _create_wrapped_model(func_name, annotation, field_info)
+        elif type_annotation in (str, int, float, bool, bytes, type(None)):
+            model = _create_wrapped_model(func_name, type_annotation)
             wrap_output = True
 
         # Case 4: Other class types (dataclasses, regular classes with annotations)
         else:
             type_hints = get_type_hints(type_annotation)
             if type_hints:
                 # Classes with type hints can be converted to Pydantic models
-                model = _create_model_from_class(type_annotation)
+                model = _create_model_from_class(type_annotation, type_hints)
```

Comment:
`_create_model_from_class()` used to compute the type hints again, pass them to avoid unnecessarily computing them again.